zingolib/src/wallet/transaction_records_by_id/trait_inputsource.rs

```diff
@@ -118,89 +114,75 @@ impl InputSource for TransactionRecordsById {
         anchor_height: zcash_primitives::consensus::BlockHeight,
         exclude: &[Self::NoteRef],
     ) -> Result<SpendableNotes<Self::NoteRef>, Self::Error> {
-        let mut sapling_note_noteref_pairs: Vec<(sapling_crypto::Note, NoteId)> = Vec::new();
-        let mut orchard_note_noteref_pairs: Vec<(orchard::Note, NoteId)> = Vec::new();
-        for transaction_record in self.values().filter(|transaction_record| {
-            transaction_record
-                .status
-                .is_confirmed_before_or_at(&anchor_height)
-        }) {
-            if sources.contains(&ShieldedProtocol::Sapling) {
-                sapling_note_noteref_pairs.extend(
-                    transaction_record
-                        .select_unspent_shnotes_and_ids::<SaplingDomain>()
-                        .into_iter()
-                        .filter(|note_ref_pair| !exclude.contains(&note_ref_pair.1)),
-                );
+        let mut unselected =
+            self.get_spendable_note_ids_and_values(sources, anchor_height, exclude);
+
+        unselected.sort_by_key(|(_id, value)| *value);
+
+        let mut selected = vec![];
+
+        let mut unselected_iterator = 0;
+        loop {
+            if unselected.is_empty() {
+                // all notes are selected. we pass the max value onwards whether we have reached target or not
+                break;
             }
-            if sources.contains(&ShieldedProtocol::Orchard) {
-                orchard_note_noteref_pairs.extend(
-                    transaction_record
-                        .select_unspent_shnotes_and_ids::<OrchardDomain>()
-                        .into_iter()
-                        .filter(|note_ref_pair| !exclude.contains(&note_ref_pair.1)),
-                );
+            match unselected.get(unselected_iterator) {
+                None => {
+                    // the iterator went off the end of the vector without finding a note big enough to complete the transaction... add the biggest note and reset the iteraton
+                    selected.push(unselected.pop().expect("nonempty"));
+                    unselected_iterator = 0;
+                    continue;
+                }
+                Some(smallest_unselected) => {
+                    // selected a note to test if it has enough value to complete the transaction on its own
+                    if smallest_unselected.1
+                        >= target_value.into_u64()
+                            - selected.iter().fold(0, |sum, (_id, value)| sum + *value)
+                    {
+                        selected.push(*smallest_unselected);
+                        unselected.remove(unselected_iterator);
```

Comment:
This line hints that `unselected_iterator` is better named `index_into_unselected`.